File: packages/desk-pilot/src/actuation.ts
/**
 * DeskPilot 执行契约：目标、手势、期望与结果。
 *
 * 设计约束（见 `.agents/specs/desk-pilot.md` D3/D4）：
 * - 目标只有两种：元素 ref（优先）或绑定 frame 的坐标（最后手段）；
 * - 每个破坏性动作都可以带 `Expectation`，动作后由 core 重新观测求值；
 * - `possiblySent` 与 `verification` 是正交的两件事：前者说"可能已经发出去了"，
 *   后者说"我们能不能证明它生效了"。两者都不允许被谎报。
 */

import type { Bounds, SurfaceRef } from "./ui-map.js";

/** 动作目标。坐标目标必须携带 frameId，否则在进入 adapter 前就被拒绝。 */
export type ActionTarget =
  | { readonly kind: "element"; readonly snapshotId: string; readonly ref: string }
  | {
      readonly kind: "coordinate";
      readonly frameId: string;
      readonly x: number;
      readonly y: number;
    };

export type PointerButton = "left" | "right" | "middle";

/** 跨平台修饰键。`cmd` 在非 darwin 上由 core 归一为 `ctrl`，模型无需分辨。 */
export type KeyModifier = "cmd" | "ctrl" | "alt" | "shift" | "win";

export interface KeyChord {
  readonly key: string;
  readonly modifiers: readonly KeyModifier[];
}

/** 指针手势。滚动没有可达性原语，走 raw 时必须让模型知道。 */
export type PointerGesture =
  | {
      readonly kind: "click";
      readonly button: PointerButton;
      readonly clicks: 1 | 2 | 3;
      readonly modifiers?: readonly KeyModifier[];
    }
  | {
      readonly kind: "drag";
      readonly from: Bounds;
      readonly to: Bounds;
      readonly button: PointerButton;
    }
  | {
      readonly kind: "scroll";
      readonly deltaX: number;
      readonly deltaY: number;
    };

export interface TextRange {
  readonly start: number;
  readonly length: number;
}

/** 文本类操作。`set` 只对 `settable` 元素合法，`select` 只对 `selectable` 合法。 */
export type TextOp =
  | { readonly kind: "type"; readonly text: string }
  | { readonly kind: "set"; readonly value: string }
  | { readonly kind: "select"; readonly range: TextRange }
  | { readonly kind: "paste"; readonly text: string };

/** 动作后期望。谓词在动作后由 core 重新观测求值。 */
export type Expectation =
  | { readonly predicate: "element_gone"; readonly target: ActionTarget }
  | { readonly predicate: "element_present"; readonly target: ActionTarget }
  | { readonly predicate: "value_is"; readonly target: ActionTarget; readonly value: string }
  | {
      readonly predicate: "surface_title_is";
      readonly surface: SurfaceRef;
      readonly title: string;
    };

/** 破坏性动作的统一信封。 */
export interface ActuationEnvelope {
  readonly target: ActionTarget;
  readonly expect?: Expectation;
  /** 调用方提供的幂等键。同一键的重复提交由 core 去重，不重复下达平台。 */
  readonly idempotencyKey?: string;
}

/** 语义动作请求：执行元素自报的 action。 */
export interface SemanticRequest extends ActuationEnvelope {
  readonly action: string;
}

/** 指针请求。 */
export interface PointerRequest extends ActuationEnvelope {
  readonly gesture: PointerGesture;
}

/** 文本请求。 */
export interface TextRequest extends ActuationEnvelope {
  readonly op: TextOp;
}

/** 执行结果的三态验证结论。`unverified` 是合法答案，不是失败。 */
export type Verification =
  | { readonly state: "verified"; readonly evidence: string }
  | { readonly state: "deviation"; readonly expected: string; readonly observed: string }
  | { readonly state: "unverified"; readonly reason: string };

/** adapter 执行后的最小回报。 */
export interface ActuationOutcome {
  /** 指令是否已经下达到平台。参数校验失败必须是 false。 */
  readonly possiblySent: boolean;
  readonly verification: Verification;
  /** 动作后产生的 snapshotId；未重新观测时为 null。 */
  readonly snapshotId: string | null;
}

/** 剪贴板操作。Wayland 上读可能不可用，此时 `text` 为 null 且必须说明原因。 */
export type ClipboardOp =
  | { readonly kind: "read" }
  | { readonly kind: "write"; readonly text: string };

export interface ClipboardResult {
  readonly text: string | null;
  readonly written: boolean;
  /** 读不可用或未授权时的原因；成功时为 null。 */
  readonly note: string | null;
}

/** `attach_cdp` 的入参与产出。 */
export interface CdpAttachRequest {
  readonly pid: number;
}

export interface CdpEndpoint {
  readonly port: number;
  readonly httpEndpoint: string;
  readonly websocketUrl: string;
  readonly product: string;
}

File: packages/desk-pilot/native/crates/surface-contract/src/actuation.rs
//! 执行侧契约：目标、手势、期望与结果。
//!
//! 与 `packages/desk-pilot/src/actuation.ts` 对应。

use serde::{Deserialize, Serialize};

use crate::ui_map::{Bounds, SurfaceRef};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum ActionTarget {
    Element { snapshot_id: String, ref_: String },
    /// 坐标目标。`frame_id` 为空即非法，host 侧在进入 daemon 前就会拒绝。
    Coordinate { frame_id: String, x: f64, y: f64 },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PointerButton {
    Left,
    Right,
    Middle,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum KeyModifier {
    Cmd,
    Ctrl,
    Alt,
    Shift,
    Win,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct KeyChord {
    pub key: String,
    #[serde(default)]
    pub modifiers: Vec<KeyModifier>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum PointerGesture {
    Click {
        button: PointerButton,
        clicks: u8,
        #[serde(default)]
        modifiers: Vec<KeyModifier>,
    },
    Drag {
        from: Bounds,
        to: Bounds,
        button: PointerButton,
    },
    Scroll { delta_x: f64, delta_y: f64 },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct TextRange {
    pub start: u32,
    pub length: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum TextOp {
    Type { text: String },
    Set { value: String },
    Select { range: TextRange },
    Paste { text: String },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "predicate", rename_all = "snake_case")]
pub enum Expectation {
    ElementGone { target: ActionTarget },
    ElementPresent { target: ActionTarget },
    ValueIs { target: ActionTarget, value: String },
    SurfaceTitleIs { surface: SurfaceRef, text: String },
}

/// 破坏性动作的统一信封。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ActuationEnvelope {
    pub target: ActionTarget,
    #[serde(default)]
    pub expect: Option<Expectation>,
    #[serde(default)]
    pub idempotency_key: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SemanticRequest {
    #[serde(flatten)]
    pub envelope: ActuationEnvelope,
    pub action: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PointerRequest {
    #[serde(flatten)]
    pub envelope: ActuationEnvelope,
    pub gesture: PointerGesture,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TextRequest {
    #[serde(flatten)]
    pub envelope: ActuationEnvelope,
    pub op: TextOp,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "state", rename_all = "snake_case")]
pub enum Verification {
    Verified { evidence: String },
    Deviation { expected: String, observed: String },
    Unverified { reason: String },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ActuationOutcome {
    /// 指令是否已下达到平台。参数校验失败必须是 false。
    pub possibly_sent: bool,
    pub verification: Verification,
    pub snapshot_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum ClipboardOp {
    Read,
    Write { text: String },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ClipboardResult {
    pub text: Option<String>,
    pub written: bool,
    pub note: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CdpAttachRequest {
    pub pid: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CdpEndpoint {
    pub port: u16,
    pub http_endpoint: String,
    pub websocket_url: String,
    pub product: String,
}
